src/main/java/com/nha/abdm/wrapper/hip/hrp/consent/ConsentService.java

```diff
@@ -1,24 +1,121 @@
 import org.springframework.stereotype.Service;
+import reactor.core.Exceptions;
 
 @Service
 public class ConsentService implements ConsentInterface {
-
-  @Autowired RequestManager requestManager;
+  private static final Logger log = LogManager.getLogger(ConsentService.class);
+  private ErrorResponse errorResponse = new ErrorResponse();
+  private final RequestManager requestManager;
+  private final HIPClient hipClient;
+  @Autowired RequestLogService requestLogService;
+  @Autowired PatientService patientService;
 
   @Value("${consentHipOnNotifyPath}")
   private String consentHipOnNotifyPath;
 
-  @Override
-  public ResponseEntity<GatewayGenericResponse> hipOnNotify(HIPOnNotifyRequest hipOnNotifyRequest) {
-    return requestManager.fetchResponseFromGateway(consentHipOnNotifyPath, hipOnNotifyRequest);
+  public ConsentService(RequestManager requestManager, HIPClient hipClient) {
+    this.requestManager = requestManager;
+    this.hipClient = hipClient;
+  }
+
+  /**
+   * The callback from ABDM gateway after consentGrant for dataTransfer, POST method for /on-notify
+   *
+   * @param hipNotifyRequest careContext and demographics details are provided, and implement a
+   *     logic to check the existence of the careContexts.
+   */
+  public void notifyOnReceived(HIPNotifyRequest hipNotifyRequest) throws IllegalDataStateException {
+    HIPOnNotifyRequest hipOnNotifyRequest = null;
+    if (hipNotifyRequest != null
+        && hipNotifyRequest.getNotification() != null
+        && hipNotifyRequest.getNotification().getConsentDetail() != null
+        && hipNotifyRequest.getNotification().getConsentDetail().getPatient() != null) {
+      HIPNotification hipNotification = hipNotifyRequest.getNotification();
+
+      RespRequest responseRequestId =
+          RespRequest.builder().requestId(hipNotifyRequest.getRequestId()).build();
+
+      if (patientService.isCareContextPresent(
+          hipNotification.getConsentDetail().getCareContexts())) {
+        Consent consent =
+            Consent.builder()
+                .status(hipNotification.getStatus())
+                .consentDetail(hipNotification.getConsentDetail())
+                .signature(hipNotification.getSignature())
+                .build();
+        patientService.addConsent(hipNotification.getConsentDetail().getPatient().getId(), consent);
+        ConsentAcknowledgement dataAcknowledgement =
+            ConsentAcknowledgement.builder()
+                .status("OK")
+                .consentId(hipNotifyRequest.getNotification().getConsentId())
+                .build();
+        hipOnNotifyRequest =
+            HIPOnNotifyRequest.builder()
+                .requestId(UUID.randomUUID().toString())
+                .timestamp(Utils.getCurrentTimeStamp())
+                .acknowledgement(dataAcknowledgement)
+                .resp(responseRequestId)
+                .build();
+      } else {
+        errorResponse.setMessage("HIP -> Mismatch of careContext");
```

Comment:
can we give more meaningful error message like the care contexts provided : <carecontexts> do not match with the care contexts present in hip wrapper database for the given patient